pins.go

```diff
@@ -0,0 +1,9 @@
+package drivers
+
+// PinOutput is hardware abstraction for a pin which outputs a
+// digital signal (high or low voltage).
+type PinOutput func(level bool)
```

Comment:
```suggestion
type PinSet func(level bool)
```
just my recent experience, this would be more intuitive when writing a new driver, same for PinInput() --> PinGet()

Reply:
At a glance PinSet conceptually sounds further from a HAL and closer to a struct field setting functionality, same for PinGet. Also PinSet sounds like it could be a set data structure.

I'd also feel odd naming them this way in Go. The "Get" and "Set" functionality is encoded in the concrete type `func(bool)` or `func() bool`. We don't call our interfaces `WriterInterface` in Go.

This is just my gut reaction speaking though.

Reply:
You are right, in English the noun for "data set" is often shorten as "set" and so can be confused with the verb. I mostly use the extension "-Func" in my own code in this case but never made an external interface with a function until now. Maybe someone has a better idea for the name. Also, because the function is located in the drivers package we are forced to repeat the "Pin" in the name. Normally such things would be located in a pin-package and a "pin.Set" would be ok.

I also understand your example with the "WriterInterface", but please have a look to my comment for the legacy interface names. The interface name can have an "-er" extension (to construct an agent noun), but the function names inside not, of course. In our case the interface name and the function name are merged to a single interface function. Maybe this idea helps to find a better name.

I just tried to use it at the opposite side - device driver side - and it felt bumpy when programming and reading at this side. I had to read the type multiple times and also the example in apa102 (construct the  softspi) to train my brain to this wording.

Reply:
What about this names, very obvious inspired by the comment: "PinSetLevel(newLevel bool)", "PinGetLevel() bool"

Reply:
So if I'm getting this correctly the name "PinOutput" is hard to understand when reading. May I suggest a couple of approaches:
- Use meaningful and semantic identifiers: Types are only part of the story. You write "PinSetLevel(newLevel bool)" as the ideal legibility standard- this is achievable! Use the variable identifier to your advantage! Variable identifiers appear in driver code much more often than the type itself and can make for extremely readable code. I suggest something along the lines of:

```go
type Device struct {
    csPinSet drivers.PinOutput
}

func (d *Device) Send(newLevel bool) {
    d.csPinSet(newLevel)
}
``` 
The actual type `drivers.PinOutput` would appear about twice in a typical driver. In the struct decvlaration and in the New function- and it appears only once per driver in the refactored version in this PR since we are trying to preserve backward compatibility with `legacy.Pin*` types. This naturally makes it very hard to adapt to an unfamiliar API since you really are not reading it at all when working with code that uses it.

- Maybe try an interface-centric approach? Change types to `PinOutputter` and `PinInputter`. See if this helps reading the code. Function handles are very similar to single-method interfaces after all. 

- Try using functions as values in other projects. Go code typically does not make use of functions as values so naturally reading code that does this feels off or unexpected when we follow a variable with parentheses. I suspect this might be a big part of the offputting part. If this is the case the name of the type is not going to aid much since the usage is far from the type's declaration in driver development. Like I mentioned earlier, use meaningful identifiers that help you read the code! If `Set` and `Get` help you read the code, then name your variables accordingly!

Reply:
Of course - please also read my other suggestions and you will find some of your ideas. "Rewording" in the Devices struct code is one of my suggestions.

And yes, "drivers.PinOutput"/"drivers.PinInput" would occur in 2 places normally - so it is maybe not worth to discuss a name change to e.g. "PinSetLevel"/"PinGetLevel" - but it is better to discuss it now, than after it is widely used. Don't get me wrong, I just try to show the confusion of a new driver developer - I personally can de-confuse and handle this by some research in the code. And, it seems nobody else had an objection about this - so please don't let my comment slow you down too much.